### PR TITLE
add deformable conv v1 op and cpu version of deformable conv v2

### PR DESCRIPTION
**Pull request by cjt222**
Deformable convolution v1 is a new method of which convolution, which add offset to every kernel to make reception field bigger and learn to capture irregular object，which is always used in object convolution and segmentation.
Inside, cpu version and gpu version of deformable convolution v1 and cpu version of Deformable convolution v2 are added. Deformable convolution v1 and v2 is used same python API, which decide which version to used by one parameter named modulated.

### CODE REVIEW

File: paddle/fluid/operators/deformable_conv_op.cc [hunk 1]
```diff
@@ -276,3 +277,12 @@ REGISTER_OPERATOR(deformable_conv, ops::DeformableConvOp,
                   ops::DeformableConvGradOpDescMaker);
 
 REGISTER_OPERATOR(deformable_conv_grad, ops::DeformableConvGradOp);
+
+using CPU = paddle::platform::CPUDeviceContext;
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_op.cc]: CPU -> CPUDeviceContext

File: paddle/fluid/operators/deformable_conv_op.h [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+#include "paddle/fluid/operators/math/blas.h"
+#include "paddle/fluid/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+template <typename T>
+T DmcnIm2colBilinearCpu(const T* bottom_data, const int data_width,
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_op.h]: 用static修饰，其他函数同

File: paddle/fluid/operators/deformable_conv_op.h [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+void FilterGradAddupCpuKernel(const int nthreads, const int n, const int height,
+                              const int width, const T* dweight_3d,
+                              T* filter_grad) {
+  for (size_t i = 0; i < nthreads; i++) {
+    filter_grad[i] = filter_grad[i] + dweight_3d[i];
+  }
+}
+
+template <typename DeviceContext, typename T>
+class DeformableConvCPUKernel : public framework::OpKernel<T> {
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_op.h:312]: 确认是CPU kernel的话就不要带typename DeviceContext

File: paddle/fluid/operators/deformable_conv_v1_op.cu [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+static constexpr int kNumCUDAThread = 512;
+static constexpr int kNumMaximumNumBlock = 4096;
+
+static inline int NumBlock(const int N) {
+  return std::min((N + kNumCUDAThread - 1) / kNumCUDAThread,
+                  kNumMaximumNumBlock);
+}
+
+template <typename T>
+__global__ void DeformableCol2imGpuKernel(
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_v1_op.cu]: Gpu -> CUDA

File: paddle/fluid/operators/deformable_conv_v1_op.cu [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+      reinterpret_cast<const platform::CUDADeviceContext&>(ctx).stream()>>>(
+      num_kernels, data_im, data_offset, im_shape[1], im_shape[2],
+      filter_shape[2], filter_shape[3], paddings[0], paddings[1], strides[0],
+      strides[1], dilations[0], dilations[1], channel_per_deformable_group,
+      col_shape[1], im_shape[0], deformable_groups, col_shape[2], col_shape[3],
+      data_col);
+}
+
+template <typename DeviceContext, typename T>
+class DeformableConvV1CUDAKernel : public framework::OpKernel<T> {
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_v1_op.cu:308]: CUDA kernel的话不要带typename DeviceContext

File: paddle/fluid/operators/deformable_conv_v1_op.h [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+          width, bp_dir);
+      val += weight * data_col_ptr[col_pos];
+      cnt += 1;
+    }
+    grad_offset[i] = val;
+  }
+}
+
+template <typename T>
+inline void DeformableCol2imCoordCpu(
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_v1_op.h]: Cpu -> CPU 其他一样

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -12022,11 +12025,20 @@ def deformable_conv(input,
...
         .. code-block:: python
 
-          data = fluid.layers.data(name='data', shape=[3, 32, 32], dtype='float32')
-          offset = fluid.layers.data(name='offset', shape=[18, 32, 32], dtype='float32')
-          mask = fluid.layers.data(name='mask', shape=[9, 32, 32], dtype='float32')
-          out = fluid.layers.deformable_conv(input=data, offset=offset, mask=mask,
-                                             num_filters=2, filter_size=3, padding=1)
+          deformable conv v2:
+         
+              data = fluid.layers.data(name='data', shape=[3, 32, 32], dtype='float32')
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py]: 这是在python的code-block，deformable conv v2:前面加#  ，下面代码缩进不要变

File: paddle/fluid/operators/deformable_conv.cu.h [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  if (h_low >= 0 && w_low >= 0) v1 = bottom_data[h_low * data_width + w_low];
+  T v2 = 0;
+  if (h_low >= 0 && w_high <= width - 1)
+    v2 = bottom_data[h_low * data_width + w_high];
+  T v3 = 0;
+  if (h_high <= height - 1 && w_low >= 0)
+    v3 = bottom_data[h_high * data_width + w_low];
+  T v4 = 0;
+  if (h_high <= height - 1 && w_high <= width - 1)
+    v4 = bottom_data[h_high * data_width + w_high];
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv.cu.h]: 这里的改成用三目运算符 ? : 的吧，会清晰一些

File: paddle/fluid/operators/deformable_conv.cu.h [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+  T w1 = hh * hw, w2 = hh * lw, w3 = lh * hw, w4 = lh * lw;
+
+  T val = (w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4);
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv.cu.h]: 不用加括号

File: paddle/fluid/operators/deformable_conv.cu.h [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  return val;
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv.cu.h]: retrun w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4

File: paddle/fluid/operators/deformable_conv.cu.h [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  if (h_low >= 0 && w_high <= width - 1)
+    v2 = bottom_data[h_low * data_width + w_high];
+  T v3 = 0;
+  if (h_high <= height - 1 && w_low >= 0)
+    v3 = bottom_data[h_high * data_width + w_low];
+  T v4 = 0;
+  if (h_high <= height - 1 && w_high <= width - 1)
+    v4 = bottom_data[h_high * data_width + w_high];
+
+  T w1 = hh * hw, w2 = hh * lw, w3 = lh * hw, w4 = lh * lw;
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv.cu.h]: 不要写在一行

File: paddle/fluid/operators/deformable_conv_op.h (hunk 2 above)
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_op.h]: 这个函数和上面的函数好像没有区别，用HOSTDEVICE修饰可以同时在CPU和CUDA上使用，不用定义两个

File: paddle/fluid/operators/deformable_conv.cu.h (hunk 11 above)
Comment (cjt222) [reply paddle/fluid/operators/deformable_conv.cu.h]: done

File: paddle/fluid/operators/deformable_conv_op.h (hunk 3 above)
Comment (cjt222) [reply paddle/fluid/operators/deformable_conv_op.h:312]: 已修复

File: paddle/fluid/operators/deformable_conv_v1_op.cu (hunk 5 above)
Comment (cjt222) [reply paddle/fluid/operators/deformable_conv_v1_op.cu:308]: 已修复

File: paddle/fluid/operators/deformable_conv.cu.h (hunk 11 above)
Comment (cjt222) [reply paddle/fluid/operators/deformable_conv.cu.h]: 已修复

**Review by heavengate — DISMISSED** [review]

**Pull request closed without merging**